### PR TITLE
reseed all Generators in Dataloader's _worker_loop()

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR addresses https://fb.workplace.com/groups/pytorch.oss.dev/posts/1699944830430051 and does a bunch of stacked changes:

1. Make `Generator` weakref-able (C++ part)
2. Create a registry of manually-created `Generator` objects (via weakrefs)
3. Use this registry in the `Dataloader`'s `_worker_loop` to re-seed all existing `Generator` instances: this extends what is already applied to the global `Generator`, which is already re-seeded.

TODO: a bit of docs and justification, which I'll do if this PR is mergeable.

CC @albanD as previously discussed

cc @ezyang @gchanan @SsnL @VitalyFedyunin @ejguan @dzhulgakov @pbelevich

### CODE REVIEW

File: torch/utils/data/_utils/worker.py [hunk 1]
```diff
@@ -223,6 +223,14 @@ def _worker_loop(dataset_kind, dataset, index_queue, data_queue, done_event,
...
         random.seed(seed)
         torch.manual_seed(seed)
+        for generator in torch.random._generator_registry:
+            # We have to reseed non-global Generators as well (TODO: explain why, will do if PR is likely to be merged)
+            # Note that we can't use generator.manual_seed(seed) here, otherwise the RNG of the generator
+            # would be the same as the global one.
+            # We also need to create a generator_seed that depends on the current generator state, otherwise
+            # all Generator instances within a given worker would have the same RNG.
+            generator_seed = torch.empty((), dtype=torch.int64).random_(generator=generator).item() + seed
+            generator.manual_seed(generator_seed)
```
Comment (NicolasHug) [inline torch/utils/data/_utils/worker.py:233]: Technically this is a change of behaviour: In `main`, the `Generator` instances have the same RNG across workers; now they don't.

I don't think users are / should rely on this behaviour anyway? The only use I can think of where users would *want* to have the same RNG across workers is if they're using the Generator to shuffle the datasets: when shuffling, you want all workers to shuffle in the same way. But for those who need that, I think it's fair to say that they should be relying on `worker_init_fn` anyway.

Comment (albanD) [reply torch/utils/data/_utils/worker.py:233]: This is technically BC-breaking so we should properly document it (please add a small paragraph in the description for the release notes). But that's ok to do it I think yes.

File: test/test_dataloader.py [hunk 2]
```diff
@@ -1563,6 +1563,39 @@ def test_multi_epochs_reproducibility(self):
...
+                return from_global, from_g1, from_g2
+
+
+        num_workers = 2
+        dataset = MyDataset(size=10, num_workers=num_workers, batch_size=1)
+        dl = DataLoader(dataset, num_workers=num_workers)
+
+        for from_global, from_g1, from_g2 in dl:
+            # Assert RNG of all Generators are different within a given worker (each "batch" comes from a single worker)
+            assert len(set([from_global, from_g1, from_g2])) == 3
```
Comment (albanD) [inline test/test_dataloader.py:1590]: Can you please remove all the plain assert from tests. You can use `self.assertEqual()`, `self.assertNot()`, etc as appropriate

File: torch/csrc/Generator.cpp [hunk 3]
```diff
@@ -79,6 +83,11 @@ static PyObject* THPGenerator_pynew(
         c10::DeviceTypeName(device.type()),
         " is not supported for torch.Generator() api.");
   }
+  self->weakreflist = NULL;
+
+  static py::handle _generator_registry = py::module::import("torch").attr("random").attr("_generator_registry");
+  _generator_registry.attr("add")(py::cast<py::object>((PyObject*)self.get()));
```
Comment (albanD) [inline torch/csrc/Generator.cpp:89]: move the add above. You don't want to do the "add" dynamically every time right?

File: torch/csrc/Generator.cpp [hunk 4]
```diff
@@ -79,6 +83,11 @@ static PyObject* THPGenerator_pynew(
         c10::DeviceTypeName(device.type()),
         " is not supported for torch.Generator() api.");
   }
+  self->weakreflist = NULL;
+
+  static py::handle _generator_registry = py::module::import("torch").attr("random").attr("_generator_registry");
```
Comment (albanD) [inline torch/csrc/Generator.cpp:88]: You should have a "release()" at the end here to make sure to leak the reference.

File: torch/random.py [hunk 5]
```diff
@@ -167,3 +168,7 @@ def fork_rng(devices=None, enabled=True, _caller="fork_rng", _devices_kw="device
         torch.set_rng_state(cpu_rng_state)
         for device, device_rng_state in zip(devices, device_rng_states):
             device_mod.set_rng_state(device_rng_state, device)
+
+# We keep track of all Generator instances (except the default one) via a registry of weak references.
```
Comment (albanD) [inline torch/random.py:172]: Ho why not the default one? Because that is the global rng and so it is already handled?

File: torch/utils/data/_utils/worker.py [hunk 6]
```diff
@@ -223,6 +223,14 @@ def _worker_loop(dataset_kind, dataset, index_queue, data_queue, done_event,
         seed = base_seed + worker_id
         random.seed(seed)
         torch.manual_seed(seed)
+        for generator in torch.random._generator_registry:
+            # We have to reseed non-global Generators as well (TODO: explain why, will do if PR is likely to be merged)
+            # Note that we can't use generator.manual_seed(seed) here, otherwise the RNG of the generator
+            # would be the same as the global one.
+            # We also need to create a generator_seed that depends on the current generator state, otherwise
+            # all Generator instances within a given worker would have the same RNG.
+            generator_seed = torch.empty((), dtype=torch.int64).random_(generator=generator).item() + seed
```
Comment (albanD) [inline torch/utils/data/_utils/worker.py:232]: Why not use `generator.initial_seed()` ?

**Review by albanD — COMMENTED** [review]
Small things but the approach sounds good to me.
Will let @ezyang give his opinion as well

**Comment by ezyang** [comment]
As we discussed earlier, I am begrudgingly ok with this approach. There are some implementation detail's I'll talk about tomorrow; one in particular is having the registry in C++ and thread safe

File: torch/utils/data/_utils/worker.py [hunk 7]
```diff
@@ -223,6 +223,14 @@ def _worker_loop(dataset_kind, dataset, index_queue, data_queue, done_event,
         seed = base_seed + worker_id
         random.seed(seed)
         torch.manual_seed(seed)
+        for generator in torch.random._generator_registry:
```
Comment (ezyang) [inline torch/utils/data/_utils/worker.py:226]: It occurs to me, technically, you don't even need the generator registry; we could just gc.get_objects() and traverse the entire live heap to look for generators 😆

Comment (ezyang) [reply torch/utils/data/_utils/worker.py:226]: Actually, legit question, why don't we do this? The patch becomes super simple then.

Comment (NicolasHug) [reply torch/utils/data/_utils/worker.py:226]: I gave it a try in https://github.com/pytorch/pytorch/pull/107131, LMK what you think

Comment (albanD) [reply torch/utils/data/_utils/worker.py:226]: Ho yeah that definitely makes it a a lot less invasive if we're ok with traversing all the alive objects on process creation. Which I think we are.

**Comment by NicolasHug** [comment]
Superseded by https://github.com/pytorch/pytorch/pull/107131
